### PR TITLE
Periodical ping

### PR DESCRIPTION
**Pull request by kpp**

### CODE REVIEW

File: examples/dht_node_new.rs [hunk 1]
```diff
@@ -168,10 +167,10 @@ fn main() {
         .map_err(|_err| Error::new(ErrorKind::Other, "Nodes timer error"));
 
     // 3 seconds for NatPingRequest
```
Comment (kpp) [inline examples/dht_node_new.rs]: Fix comment

Comment (NamsooCho) [reply examples/dht_node_new.rs]: I will fit it on next commit.

**Comment by kpp** [comment]
Pls rebase on master

**Comment by kpp** [comment]
There should be no 2 commits with the same actions be in one PR:
```
->  refactor(dht): move rand and tokio-timer crate position
    refactor(dht): PingRequest/Response & NodesRequest/Response are working
->  refactor(dht): move rand and tokio-timer position 
```

File: examples/dht_node.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    let server_obj_c = server_obj.clone();
+    let ping_sender = ping_wakeups.for_each(move |()| {
+            // println!("ping wakeup");
+            let kbucket_c = server_obj_c.borrow().kbucket.clone().iter();
+            let server_obj_c_c = server_obj_c.clone();
+
+            type LoopFuture = Box<Future<Item = future::Loop<KbucketIter, KbucketIter>, Error = ()>>;
+
+            let packet_sender = future::loop_fn(kbucket_c, move |mut iter| -> LoopFuture {
```
Comment (kpp) [inline examples/dht_node.rs]: Try to use https://github.com/jgallagher/tokio-chat-example/blob/master/tokio-chat-server/src/main.rs#L86..L105

File: examples/dht_node.rs [hunk 3]
```diff
@@ -128,28 +126,16 @@ fn main() {
...
-            })
-            .map(|_| ());
-
-            handle.spawn(packet_sender);
-            Ok(())
+        // println!("ping wakeup");
+        let kbucket_c = server_obj_c.borrow().kbucket.clone().iter();
+        let server_obj_c_c = server_obj_c.clone();
+
+        let all_sends = kbucket_c.map(|peer|
```
Comment (kpp) [inline examples/dht_node.rs]: good job! Now rename it into `Server::send_pings` that will return a future that sends pings to all it's clients

File: src/lib.rs [hunk 4]
```diff
@@ -58,13 +58,14 @@ extern crate quickcheck;
 extern crate tokio;
 extern crate tokio_core;
 extern crate tokio_io;
+extern crate rand;
+extern crate tokio_timer;
```
Comment (kpp) [inline src/lib.rs]: We don't use it in our code anymore, do we?

File: src/lib.rs [hunk 5]
```diff
@@ -58,13 +58,14 @@ extern crate quickcheck;
 extern crate tokio;
 extern crate tokio_core;
 extern crate tokio_io;
+extern crate rand;
```
Comment (kpp) [inline src/lib.rs]:
```
#[cfg(test)]
extern crate rand;
```

File: examples/dht_node.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        // println!("ping wakeup");
+        let kbucket_c = server_obj_c.borrow().kbucket.clone().iter();
+        let server_obj_c_c = server_obj_c.clone();
+
+        let send_pings = kbucket_c.map(|peer|
+            server_obj_c_c.borrow_mut().send_ping_req(peer));
+
+        let pings_stream = stream::futures_unordered(send_pings).then(|_| Ok(()));
+
+        pings_stream.for_each(|()| Ok(()))
```
Comment (kpp) [inline examples/dht_node.rs]: Indent.

File: examples/dht_node.rs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            err
+        });
+
+    // 60 seconds for PingRequests
+    let ping_wakeups = Interval::new(Duration::from_secs(60));
+
+    let server_obj_c = server_obj.clone();
+    let ping_sender = ping_wakeups.for_each(move |()| {
+        // println!("ping wakeup");
+        let kbucket_c = server_obj_c.borrow().kbucket.clone().iter();
```
Comment (kpp) [inline examples/dht_node.rs]: Now rename this block of code and move it into `Server::send_pings` that will return a future that sends pings to all it's clients

File: src/toxcore/dht/kbucket.rs [hunk 8]
```diff
@@ -354,6 +355,18 @@ impl Kbucket {
...
     }
 
+    /// get random node to select peer to send NodesRequest
+    pub fn get_random_node(&self) -> Option<PackedNode> {
+        if self.is_empty() {
+            return None
+        }
+        let mut num_k: u8 = rand::thread_rng().gen_range(0, self.size());
+        while self.buckets[num_k as usize].nodes.len() == 0 {
+            num_k = rand::thread_rng().gen_range(0, self.size());
```
Comment (kpp) [inline src/toxcore/dht/kbucket.rs]: Maybe better use `random_u64` or `random_u32` from `tox::toxcore::crypto_core`?

File: src/toxcore/dht/server.rs [hunk 9]
```diff
@@ -576,19 +662,18 @@ mod tests {
...
-        assert!(!alice.handle_nodes_req(client, nodes_req).wait().is_ok());
+        let client = alice.create_client(&addr, bob_pk);
+        assert!(alice.handle_nodes_req(client, nodes_req).wait().is_err());
     }
 
     // handle_nodes_resp()
     #[test]
     fn server_handle_nodes_resp_test() {
-        let (mut alice, precomp, bob_pk, _rx, addr) = create_node();
+        let (mut alice, precomp, bob_pk, _, _rx, addr) = create_node();
```
Comment (kpp) [inline src/toxcore/dht/server.rs]: `_` - what does it mean?

File: src/toxcore/dht/server.rs [hunk 10]
```diff
@@ -637,21 +732,153 @@ mod tests {
...
+        let alice_pk = alice.pk.clone();
+        let client = alice.create_client(&addr.clone(), alice_pk);
+        assert!(alice.handle_nat_ping_req(client, dht_req, nat_req).wait().is_ok());
     }
 
     // handle nat ping response
     #[test]
     fn server_handle_nat_ping_resp_test() {
-        let (mut alice, precomp, bob_pk, _rx, addr) = create_node();
+        let (mut alice, precomp, bob_pk, _, _rx, addr) = create_node();
```
Comment (kpp) [inline src/toxcore/dht/server.rs]: `_`

File: src/toxcore/dht/server.rs [hunk 11]
```diff
@@ -352,35 +387,33 @@ impl Server {
...
-        }
+        crypto_init();
 
         let (pk, sk) = gen_keypair();
         let (tx, rx) = mpsc::unbounded::<(DhtPacket, SocketAddr)>();
         let alice = Server::new(tx, pk, sk);
         let (bob_pk, bob_sk) = gen_keypair();
         let precomp = precompute(&alice.pk, &bob_sk);
-
+
```
Comment (kpp) [inline src/toxcore/dht/server.rs]: remove trailing spaces

File: src/toxcore/dht/kbucket.rs [hunk 12]
```diff
@@ -354,6 +354,18 @@ impl Kbucket {
...
     }
 
+    /// get random node to select peer to send NodesRequest
+    pub fn get_random_node(&self) -> Option<PackedNode> {
+        if self.is_empty() {
+            return None
+        }
+        let mut num_k = random_u32() % self.size() as u32;
+        while self.buckets[num_k as usize].nodes.len() == 0 {
+            num_k = random_u32() % self.size() as u32;
```
Comment (kpp) [inline src/toxcore/dht/kbucket.rs]: You may use `random_u64` not to cast it as `u32`, because usize=u64

**Pull request merged**